compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2690,29 +2694,110 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, String> {
+                return;
             }
-        } else {
-            sess.fatal("option `-Z gcc-ld` is used even though linker flavor is not gcc");
         }
     }
+
+    // From now on in this function, we handle the `gcc` linker-flavor enrichment to use `lld`.
+
+    // Start by checking if we're in the context of a known issue that users might hit when
+    // using `lld`:
+    //
+    // 1. when requesting self-contained CRT linking (or on a target that does it
+    // automatically), and coverage/profile generation: point at #79555 "Coverage is not
+    // generated when using rust-lld as linker"
+    let instrument_coverage =
+        cg.instrument_coverage.is_some() && cg.instrument_coverage != Some(InstrumentCoverage::Off);
+    let generate_profile = cg.profile_generate.enabled();
+    if crt_objects_fallback && (instrument_coverage || generate_profile) {
+        sess.warn(
+            "Using `lld`, self-contained linking, and coverage or profile generation has known \
+            issues. See issue #79555 for more details, at \
+            https://github.com/rust-lang/rust/issues/79555",
+        );
+    }
+
+    // 2. Maybe point at https://github.com/flamegraph-rs/flamegraph/pull/157 or the
+    // corresponding rust/LLVM issue when/if it's tracked, depending on whether we use the
+    // workaround argument `--no-rosegment` by default when invoking `lld`.
+    //
+    // 3. If in the future, other linker flavors and targets are eligible to a `rust-lld`
+    // enrichment, maybe also point at target-specific issues like:
+    // - MSVC + ThinLTO blocker https://github.com/rust-lang/rust/issues/81408
+    // - the "lld on MSVC" tracking issue https://github.com/rust-lang/rust/issues/71520
+    //   containing a list of blocking issues
+
+    // Now, handle `rust-lld`. If the `-Zgcc-ld=lld` flag was provided, we use `rust-lld`, the
+    // rustup-distributed version of `lld` (when applicable, i.e. not in distro-builds) by:
+    // - checking the `lld-wrapper`s exist in the sysroot
+    // - adding their folder as a search path, or requesting to use a wrapper directly
+    if unstable_gcc_lld {
+        // A `gcc-ld` folder (containing the `lld-wrapper`s that will run `rust-lld`) is present in
+        // the sysroot's target-specific tool binaries folder.
+        let tools_path = sess.get_tools_search_paths(false);
+        let gcc_ld_dir = tools_path
+            .into_iter()
+            .map(|p| p.join("gcc-ld"))
+            .find(|p| p.join(if sess.host.is_like_windows { "ld.exe" } else { "ld" }).exists())
+            .unwrap_or_else(|| sess.fatal("rust-lld (as ld) not found"));
+
+        cmd.arg({
+            let mut arg = OsString::from("-B");
+            arg.push(gcc_ld_dir);
+            arg
+        });
+        cmd.arg(format!("-Wl,-rustc-lld-flavor={}", sess.target.lld_flavor.as_str()));
+    } else {
+        // Otherwise, we were asked to use `lld` but not `rust-lld`.
+        cmd.arg("-fuse-ld=lld");
```

Comment:
Maybe this PR should wait for https://github.com/rust-lang/rust/pull/100200, in addition to fixing the regression it basically makes the "self-contained" and "lld" aspects of `-Zgcc-ld` orthogonal and removes useless checks like
```rust
    // - checking the `lld-wrapper`s exist in the sysroot
```
above, that check something that is not necessarily true, and will prevent finding system lld instead.